web.config.txt
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
   <system.webServer>
       <rewrite>
           <rules>
               <rule name="Joomla! Rule 1" stopProcessing="true">
                   <match url="^(.*)$" ignoreCase="false" />
                   <conditions logicalGrouping="MatchAny">
                       <add input="{QUERY_STRING}" pattern="base64_encode[^(]*\([^)]*\)" ignoreCase="false" />
                       <add input="{QUERY_STRING}" pattern="(&gt;|%3C)([^s]*s)+cript.*(&lt;|%3E)" />
                       <add input="{QUERY_STRING}" pattern="GLOBALS(=|\[|\%[0-9A-Z]{0,2})" ignoreCase="false" />
                       <add input="{QUERY_STRING}" pattern="_REQUEST(=|\[|\%[0-9A-Z]{0,2})" ignoreCase="false" />
                   </conditions>
                   <action type="CustomResponse" url="index.php" statusCode="403" statusReason="Forbidden" statusDescription="Forbidden" />
               </rule>
               <rule name="Joomla! Rule 2">
                   <match url="(.*)" ignoreCase="false" />
                   <conditions logicalGrouping="MatchAll">
                     <add input="{URL}" pattern="^/index.php" ignoreCase="true" negate="true" />
                     <add input="{URL}" pattern="/component/|(/[^.]*|\.(php|html?|feed|pdf|vcf|raw))$" />
                     <add input="{REQUEST_FILENAME}" matchType="IsFile" ignoreCase="false" negate="true" />
                     <add input="{REQUEST_FILENAME}" matchType="IsDirectory" ignoreCase="false" negate="true" />
                   </conditions>
                   <action type="Rewrite" url="index.php" />
               </rule>
           </rules>
       </rewrite>
   </system.webServer>
</configuration>


tests/CHANGELOG
0. Metadata
-----------
Copyright:		Copyright (C) 2005 - 2012 Open Source Matters, Inc. All rights reserved.
License:		GNU General Public License version 2 or later; see LICENSE.txt

1. Copyright and disclaimer
---------------------------
This application is opensource software released under the GPL.  Please
see source code and the LICENSE file


2. Changelog
------------
This is a non-exhaustive (but still near complete) changelog for
Joomla! 1.6 automated tests, including alpha and beta versions.
Our thanks to all those people who've contributed bug reports and
code fixes.

Legend:

* -> Security Fix
# -> Bug Fix
$ -> Language fix or change
+ -> Addition
^ -> Change
- -> Removed
! -> Note

10-Feb-2012 Mark Dexter
 # Clear cache after system tests
 + Add Module0002 test

17-Jan-2012 Mark Dexter
 # Fix failing system tests for sample data changes

16-Jan-2012 Mark Dexter
 # Fix failing system test for Control Panel change

21-Dec-2011 Mark Dexter
 # Fix failing system test for sample data change

18-Dec-2011 Mark Dexter
 # Change link name for read private messages

16-Dec-2011 Mark Dexter
 # Fix failing system tests (for menu type changes in trunk)

12-Nov-2011 Mark Dexter
 + Added system test to test modals in front-end article edit

06-Nov-2011 Mark Dexter
 # Fix failing system test (for db17 branch)

25-Oct-2011 Mark Dexter
 # Fix failing system test due to module removal

26-Sep-2011 Mark Dexter
 + Added system test for path disclosure security issue
 + Added system test for installation in a non-default language

12-Sep-2011 Mark Dexter
 # Fix system test for change in sample data module positions

23-Aug-2011 Mark Dexter
 # Fix system test for h3 tag change

09-Jul-2011 Mark Dexter
 # Fix system test for changed message

06-Jul-2011 Mark Dexter
 ^ Force test db prefix to jos_ 

04-Jul-2011 Mark Dexter
 # Fix failing system test

28-Jun-2011 Mark Dexter
 # Fix system tests for issue 23125

27-Jun-2011 Mark Dexter
 + Add security0001Test to test some xss fixes

20-Jun-2011 Mark Dexter
 # Fix unit test JFilterInput to run correctly

06-Jun-2011 Mark Dexter
 # Fix doInstall for AJAX changes made to installation screens
 ^ Modify SeleniumJoomlaTestCase to check for PHP warnings as well as notices

30-May-2011 Mark Dexter
 # Fix failing system test from change in bluestork template

17-May-2011 Christophe Demko
 # [#25894] *Fatal Error: JDocumentFeed disabled (Jean-Marie Simonet)

14-May-2011 Christophe Demko
 # [#25772] Librairie JDate translates month and day strings for RFC822 format which is not correct

04-May-2011 Mark Dexter
# Fix redirect test

18-Apr-2011 Mark Dexter
+ System test for redirect component

02-Apr-2011 Mark Dexter
# Fix doInstall to set error reporting to Maximum

31-Mar-2011 Mark Dexter
+ [#22618] Automated ACL system test to test manage permission (Tim Plummer)

23-Mar-2011 Mark Dexter
# [#25359] Fix failing unit test for JDocumentRenderRSS

16-Mar-2011 Mark Dexter
# [#25286] Fix failing unit test for JDocument (Rouven Weßling)
# [#25301] Xcache can not be tested using phpunit (Rouven Weßling)
# [#25348] Add more unit tests for JURI (Rouven Weßling)

18-Feb-2011 Mark Dexter
# Fix failing system tests for "published" language change, 404 error change, Save & Close ACL modal change
# Fix failing system tests for id changes made in [#24947]

26-Jan-2011 Mark Dexter
# Fix failing system test

11-Jan-2011 Mark Dexter
# Fix problems with acl0001Test

10-Jan-2011 Mark Dexter
# Fix failing system tests

07-Jan-2011 Mark Dexter
# Fix failing system test

02-Jan-2011 Christophe Demko
 # Correcting unit test since commit #20076

03-Jan-2011 Mark Dexter
# Fix problem with unit tests with mysqli driver

02-Jan-2011 Mark Dexter
# Fixed failing system tests
# Fixed failing JUserHelperTest

02-Jan-2011 Christophe Demko
 + Added test for 3rd parameter to JRegistry::loadFile

17-Dec-2010 Ian MacLennan
 + Added test for JToolbar::addButtonPath

14-Dec-2010 Mark Dexter
# Check that menu item type displays when editing a menu item

14-Dec-2010 Christophe Demko
 # Add firstDay metadata in unit tests

11-Dec-2010 Christophe Demko
 # [#22648] function isCompatible incorrectly checks current version (Mark Dexter)

08-Dec-2010 Mark Dexter
# Fix failing sample_data0001Test.php 
^ Add option to doInstall.php to set hathor as default admin template

06-Dec-2010 Christophe Demko
 # Test coverage for the JArrayHelper::fromObject method

04-Dec-2010 Christophe Demko
 # Test coverage for automatic generated field name

30-Nov-2010 Mark Dexter
# Fix failing JStringTest and JArrayHelperTest due to locale settings
# Skip some string sort tests if MAC OS

27-Nov-2010 Mark Dexter
# Improved test coverage for JUserTest

17-Nov-2010 Mark Dexter
# Fix doInstall to allow selection of different database types
^ Added two methods to JFolderTest (Christophe Demoko) as part of [#22484]

15-Nov-2010 Mark Dexter
# Fix failing system test for sample data changes

05-Nov-2010 Mark Dexter
# Fix failing system tests for ACL language changes
# Unit test for [#22444] JRegistry fails when binding an object which contains a JRegistry object

01-Nov-2010 Mark Dexter
# Fix doInstall so it can be run against a different Joomla! instance
# Small fix for featured0001Test due to default ordering in Featured Articles Manager

30-Oct-2010 Mark Dexter
# Remove tests/system/server/configdef.php file accidentally added to version control

29-Oct-2010 Mark Dexter
# Fix problem in controlpanel004Test for small change in category ui

28-Oct-2010 Mark Dexter
# Fix bugs in menu0001Test from new template position modal

22-Oct-2010 Mark Dexter
# Fix bug in JAccessTest and JAccess class when invalid group id is checked
+ Add unit tests for JSession

14-Oct-2010 Mark Dexter

installation/models/configuration.php
<?php
/**
 * @package		Joomla.Installation
 * @copyright	Copyright (C) 2005 - 2012 Open Source Matters. All rights reserved.
 * @license		GNU General Public License version 2 or later; see LICENSE.txt
 */

defined('_JEXEC') or die;

jimport('joomla.application.component.model');
jimport('joomla.filesystem.file');
require_once JPATH_INSTALLATION.'/helpers/database.php';

/**
 * Install Configuration model for the Joomla Core Installer.
 *
 * @package		Joomla.Installation
 * @since		1.6
 */
class JInstallationModelConfiguration extends JModel
{
	/**
	 * @return boolean
	 */
	public function setup($options)
	{
		// Get the options as a JObject for easier handling.
		$options = JArrayHelper::toObject($options, 'JObject');

		// Attempt to create the root user.
		if (!$this->_createConfiguration($options)) {
			return false;
		}

		// Attempt to create the root user.
		if (!$this->_createRootUser($options)) {
			return false;
		}

		return true;
	}

	function _createConfiguration($options)
	{
		// Create a new registry to build the configuration options.
		$registry = new JRegistry();

		/* Site Settings */
		$registry->set('offline', $options->site_offline);
		$registry->set('offline_message', JText::_('INSTL_STD_OFFLINE_MSG'));
		$registry->set('display_offline_message', 1);
		$registry->set('offline_image', '');
		$registry->set('sitename', $options->site_name);
		$registry->set('editor', 'tinymce');
		$registry->set('captcha', '0');
		$registry->set('list_limit', 20);
		$registry->set('access', 1);

		/* Debug Settings */
		$registry->set('debug', 0);
		$registry->set('debug_lang', 0);

		/* Database Settings */
		$registry->set('dbtype', $options->db_type);
		$registry->set('host', $options->db_host);
		$registry->set('user', $options->db_user);
		$registry->set('password', $options->db_pass);
		$registry->set('db', $options->db_name);
		$registry->set('dbprefix', $options->db_prefix);

		/* Server Settings */
		$registry->set('live_site', '');
		$registry->set('secret', JUserHelper::genRandomPassword(16));
		$registry->set('gzip', 0);
		$registry->set('error_reporting', 'default');
		$registry->set('helpurl', 'http://help.joomla.org/proxy/index.php?option=com_help&amp;keyref=Help{major}{minor}:{keyref}');
		$registry->set('ftp_host', $options->ftp_host);
		$registry->set('ftp_port', $options->ftp_port);
		$registry->set('ftp_user', $options->ftp_save ? $options->ftp_user : '');
		$registry->set('ftp_pass', $options->ftp_save ? $options->ftp_pass : '');
		$registry->set('ftp_root', $options->ftp_save ? $options->ftp_root : '');
		$registry->set('ftp_enable', $options->ftp_enable);

		/* Locale Settings */
		$registry->set('offset', 'UTC');
		$registry->set('offset_user', 'UTC');

		/* Mail Settings */
		$registry->set('mailer', 'mail');
		$registry->set('mailfrom', $options->admin_email);
		$registry->set('fromname', $options->site_name);
		$registry->set('sendmail', '/usr/sbin/sendmail');
		$registry->set('smtpauth', 0);
		$registry->set('smtpuser', '');
		$registry->set('smtppass', '');
		$registry->set('smtphost', 'localhost');
		$registry->set('smtpsecure', 'none');
		$registry->set('smtpport', '25');

		/* Cache Settings */
		$registry->set('caching', 0);
		$registry->set('cache_handler', 'file');
		$registry->set('cachetime', 15);

		/* Meta Settings */
		$registry->set('MetaDesc', $options->site_metadesc);
		$registry->set('MetaKeys', $options->site_metakeys);
		$registry->set('MetaTitle', 1);
		$registry->set('MetaAuthor', 1);
		$registry->set('MetaVersion', 0);
		$registry->set('robots', '');

		/* SEO Settings */
		$registry->set('sef', 1);
		$registry->set('sef_rewrite', 0);
		$registry->set('sef_suffix', 0);
		$registry->set('unicodeslugs', 0);

		/* Feed Settings */
		$registry->set('feed_limit', 10);
		$registry->set('log_path', JPATH_ROOT . '/logs');
		$registry->set('tmp_path', JPATH_ROOT . '/tmp');

		/* Session Setting */
		$registry->set('lifetime', 15);
		$registry->set('session_handler', 'database');

		// Generate the configuration class string buffer.
		$buffer = $registry->toString('PHP', array('class'=>'JConfig', 'closingtag' => false));


		// Build the configuration file path.
		$path = JPATH_CONFIGURATION . '/configuration.php';

		// Determine if the configuration file path is writable.
		if (file_exists($path)) {
			$canWrite = is_writable($path);
		} else {
			$canWrite = is_writable(JPATH_CONFIGURATION . '/');
		}

		/*
		 * If the file exists but isn't writable OR if the file doesn't exist and the parent directory
		 * is not writable we need to use FTP
		 */
		$useFTP = false;
		if ((file_exists($path) && !is_writable($path)) || (!file_exists($path) && !is_writable(dirname($path).'/'))) {
			$useFTP = true;
		}

		// Check for safe mode
		if (ini_get('safe_mode')) {
			$useFTP = true;
		}

		// Enable/Disable override
		if (!isset($options->ftpEnable) || ($options->ftpEnable != 1)) {
			$useFTP = false;
		}

		if ($useFTP == true) {
			// Connect the FTP client
			jimport('joomla.client.ftp');
			jimport('joomla.filesystem.path');

			$ftp = JFTP::getInstance($options->ftp_host, $options->ftp_port);
			$ftp->login($options->ftp_user, $options->ftp_pass);

			// Translate path for the FTP account
			$file = JPath::clean(str_replace(JPATH_CONFIGURATION, $options->ftp_root, $path), '/');

			// Use FTP write buffer to file
			if (!$ftp->write($file, $buffer)) {
				// Set the config string to the session.
				$session = JFactory::getSession();
				$session->set('setup.config', $buffer);
			}

			$ftp->quit();
		} else {
			if ($canWrite) {
				file_put_contents($path, $buffer);
				$session = JFactory::getSession();
				$session->set('setup.config', null);
			} else {
				// Set the config string to the session.
				$session = JFactory::getSession();
				$session->set('setup.config', $buffer);
			}
		}

		return true;
	}

	function _createRootUser($options)
	{
		// Get a database object.
		try
		{
			$db = JInstallationHelperDatabase::getDBO($options->db_type, $options->db_host, $options->db_user, $options->db_pass, $options->db_name, $options->db_prefix);
		}
		catch (RuntimeException $e)
		{
			$this->setError(JText::sprintf('INSTL_ERROR_CONNECT_DB', $e->getMessage()));
		}

		// Create random salt/password for the admin user
		$salt = JUserHelper::genRandomPassword(32);
		$crypt = JUserHelper::getCryptedPassword($options->admin_password, $salt);
		$cryptpass = $crypt.':'.$salt;

		// create the admin user
		date_default_timezone_set('UTC');
		$installdate	= date('Y-m-d H:i:s');
		$nullDate		= $db->getNullDate();
		//sqlsrv change
		$query = $db->getQuery(true);
		$query->select('id');
		$query->from('#__users');
		$query->where('id = 42');

		$db->setQuery($query);

		if($db->loadResult())
		{
		  $query = $db->getQuery(true);
		  $query->update('#__users');
		  $query->set('name = '.$db->quote('Super User'));
		  $query->set('username = '.$db->quote($options->admin_user));
		  $query->set('email = '.$db->quote($options->admin_email));
		  $query->set('password = '.$db->quote($cryptpass));
		  $query->set('usertype = '.$db->quote('deprecated'));
		  $query->set('block = 0');
		  $query->set('sendEmail = 1');
		  $query->set('registerDate = '.$db->quote($installdate));
		  $query->set('lastvisitDate = '.$db->quote($nullDate));
		  $query->set('activation = '.$db->quote('0'));
		  $query->set('params = '.$db->quote(''));
		  $query->where('id = 42');
		} else {
		 $query = $db->getQuery(true);
		  $columns =  array($db->quoteName('id'), $db->quoteName('name'), $db->quoteName('username'),
							$db->quoteName('email'), $db->quoteName('password'),
							$db->quoteName('usertype'),
							$db->quoteName('block'),
							$db->quoteName('sendEmail'), $db->quoteName('registerDate'),
							$db->quoteName('lastvisitDate'), $db->quoteName('activation'), $db->quoteName('params'));
		  $query->insert('#__users', true);
		  $query->columns($columns);

		  $query->values('42'. ', '. $db->quote('Super User'). ', '. $db->quote($options->admin_user). ', '.
					$db->quote($options->admin_email). ', '. $db->quote($cryptpass). ', '. $db->quote('deprecated').', '.$db->quote('0').', '.$db->quote('1').', '.
					$db->quote($installdate).', '.$db->quote($nullDate).', '.$db->quote('0').', '.$db->quote(''));
	}
	$db->setQuery($query);
	if (!$db->query()) {
		$this->setError($db->getErrorMsg());
		return false;
	}

	// Map the super admin to the Super Admin Group
	$query = $db->getQuery(true);
	$query->select('user_id');
	$query->from('#__user_usergroup_map');
	$query->where('user_id = 42');

	$db->setQuery($query);

    if($db->loadResult())
    {
      $query = $db->getQuery(true);
      $query->update('#__user_usergroup_map');
      $query->set('user_id = 42');
      $query->set('group_id = 8');
    } else {
      $query = $db->getQuery(true);
      $query->insert('#__user_usergroup_map', false);
	  $query->columns(array($db->quoteName('user_id'), $db->quoteName('group_id')));
	  $query->values('42'. ', '. '8');

    }

	$db->setQuery($query);
	if (!$db->query()) {
		$this->setError($db->getErrorMsg());
		return false;
	}

	return true;
	}
}
